### ISSUE TITLE
stability: determine reason behind remaining Raft snapshots

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#8639]
#8636 eliminated the majority of Raft initiated snapshots, but a 10m `block_writer` run against a local 4 node cluster shows about 1/3 of the remaining snapshots are due to Raft:

```
~/Development/go/src/github.com/cockroachdb/cockroach pmattis/raft-log-queue egrep -h -o 'applied (preemptive|Raft) snapshot' cockroach-data/*/logs/0.stderr | sort | uniq -c
  12 applied Raft snapshot
  25 applied preemptive snapshot
```

Naively I would expect 0 Raft snapshots in this test setup.

**Comment by petermattis** [comment]
And another runs shows:

```
~/Development/go/src/github.com/cockroachdb/cockroach pmattis/raft-log-queue egrep -h -o 'applied (preemptive|Raft) snapshot' cockroach-data/*/logs/0.stderr | sort | uniq -c
  37 applied Raft snapshot
  32 applied preemptive snapshot
```

**Comment by petermattis** [comment]
Similar to #8636, we're truncating the raft log too aggressively. When we add a new replica we send a preemptive snapshot before the new replica is added to the raft group. Raft log truncation is based on the members of the Raft group, so we merrily proceed to truncate the log while the preemptive snapshot is in flight. The new replica starts up, sees that it has missed some log entries, requests them from the leader which then has to send a Raft snapshot.

**Comment by tamird** [comment]
Was this meant to be closed by #8639?

**Comment by tamird** [comment]
Ah, it's not closing because that didn't merge to the default branch.
